application-xpoll-api/src/main/java/com/xwiki/xpoll/internal/DefaultXPollManager.java

```diff
@@ -0,0 +1,203 @@
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.xpoll.internal;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReferenceSerializer;
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.model.reference.SpaceReference;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xwiki.xpoll.XPollException;
+import com.xwiki.xpoll.XPollManager;
+
+@Component
+@Singleton
+public class DefaultXPollManager implements XPollManager
+{
+    private static final String XPOLL_SPACE_NAME = "XPoll";
+
+    private static final LocalDocumentReference XPOLL_CLASS_REFERENCE =
+        new LocalDocumentReference(XPOLL_SPACE_NAME, "XPollClass");
+
+    private static final LocalDocumentReference XPOLL_VOTES_CLASS_REFERENCE =
+        new LocalDocumentReference(XPOLL_SPACE_NAME, "XPollVoteClass");
+
+    private static final String PROPOSALS = "proposals";
+
+    private static final String VOTES = "votes";
+
+    private static final String WINNER = "winner";
+
+    private static final String USER = "user";
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> serializer;
+
+    @Override
+    public void vote(DocumentReference docReference, DocumentReference user, List<String> votedProposals)
+        throws XPollException
+    {
+        XWikiContext context = contextProvider.get();
+        XWikiDocument document = null;
+        try {
+            document = context.getWiki().getDocument(docReference, context);
+            setUserVotes(votedProposals, context, document, user);
+            updateWinner(context, document);
+        } catch (XWikiException e) {
+            throw new XPollException("Failed to do a document specific operation: ", e);
```

Comment:
The exception message should match what this method does.

```suggestion
            throw new XPollException(String.format("Failed to vote for [%s] on behalf of [%s].", docReference, user), e);
```